core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java

```diff
@@ -4279,8 +4279,12 @@ private void analyzeSelectAllColumns(
                             .orElseThrow(() -> semanticException(TABLE_NOT_FOUND, allColumns, "Unable to resolve reference %s", prefix));
                     if (identifierChainBasis.getBasisType() == TABLE) {
                         RelationType relationType = identifierChainBasis.getRelationType().orElseThrow();
-                        List<Field> fields = filterInaccessibleFields(relationType.resolveVisibleFieldsWithRelationPrefix(Optional.of(prefix)));
+                        List<Field> requestedFields = relationType.resolveVisibleFieldsWithRelationPrefix(Optional.of(prefix));
+                        List<Field> fields = filterInaccessibleFields(requestedFields);
                         if (fields.isEmpty()) {
+                            if (!requestedFields.isEmpty()) {
+                                throw semanticException(TABLE_NOT_FOUND, allColumns, "Relation not found or not allowed");
```

Comment:
Or should this be AccessDeniedException?

Reply:
I went back and forth on this; I left it as not found to:

* Match current `COLUMN_NOT_FOUND` error code
* Leave room for columns to be filtered out not due to access control reasons

However I would be open to change.